rfcs/2020-06-18-2625-architecture-revisit.md

```diff
@@ -0,0 +1,504 @@
+
+As we build out this compilation functionality, we should take the opportunity
+to keep all the transformations as modular as possible. Similar to phases in
+a traditional compiler, we can take the complex user input of our TOML config
+and iteratively consolidate and lower it to a simpler representation. Each step
+can be independently written and tested, and we can even make certain
+optimizations opt-in experiments or configurable.
+
+## Prior art
+
+The idea of running full copies of the processing graph in each worker
+thread/task comes from [timely
+dataflow](https://www.youtube.com/watch?v=yOnPmVf4YWo), [Kafka
+Streams](https://kafka.apache.org/25/documentation/streams/architecture#streams_architecture_threads),
+and
+[Logstash](https://www.elastic.co/guide/en/logstash/current/execution-model.html)
+(to a lesser degree). In each case, the strategy was to increase throughput via
+concurrency enabled by partitioning. Homogeneous workers help reduce complexity
+of resource allocation, lower communication overhead, etc.
+
+Timely and Kafka Streams are both libraries rather than configurable tools, so
+there was less to draw from in terms of dynamic topology building and/or
+compilation.
+[Logstash](https://github.com/elastic/logstash/blob/78c7204552d893fe176e516cd923bf0cc1f4d052/logstash-core/src/main/java/org/logstash/config/ir/ConfigCompiler.java),
+on the other hand, is similarly configurable and has a whole compiler-like
+infrastructure with IR, etc, to build what they call "datasets" (like our idea
+of pipelines above).
+
+Some other interesting systems that I wasn't able to dig into as deeply are
+[Materialize](https://github.com/MaterializeInc/materialize) and
+[declarative-dataflow](https://github.com/comnik/declarative-dataflow). Both
+build on top of timely dataflow, but do so in a way that computation is defined
+at runtime instead of compile-time. There is likely some useful knowledge to be
+gained by investigating how they build processing topologies at runtime. On the
+other hand, both system focus significantly more on computation than Vector,
+which focuses more on data movement.
+
+## Sales Pitch
+
+The changes proposed above put us in a much stronger architectural position
+moving forward. The most important differences are focused on increasing the
+amount of information available at runtime and allowing us to make more
+intelligent decisions. Even if the specific implementation choice of pushing
+transform into source tasks does not work out, the same interface changes would
+allow alternative options (e.g. a threadpool) with no change to components
+themselves.
+
+The concept of pipelines per source task gives us a concurrency model that
+scales naturally with load while taking advantage of natural partitioning of the
+data. It can be extended later with some user-facing idea of partitioning, but
+we can start to reap the benefits of partitioning without putting that burden on
+users.
```

Comment:
Yes! From a UX perspective I do not want to concern the user with partitioning, concurrency, and so on. I lean very much towards convention over configuration for Vector.